Add cxxstd variant to ecflow, update with latest ecflow version

srherbener opened the pull request:
This PR is adding features to the ecflow package.py to help support an effort I am going through to get our JEDI spack-stack build to work on the Mac using the apple-clang@15.0.0 compiler.

The purpose of the addition of the cxxstd variant is so we can compile using the c++17 language standard, and get the BOOST_NO_CXX98_FUNCTION_BASE appropriately set for c++17 builds. See https://github.com/JCSDA/spack-stack/issues/877 for more details.

srherbener committed:
Added recent versions to ecflow/package.py, as well as added a cxxstd variant that is
needed to set BOOST_NO_CXX98_FUNCTION_BASE appropriately when building with C++17 standard.

climbfuji commented on var/spack/repos/builtin/packages/ecflow/package.py:
```diff
@@ -24,6 +24,8 @@ class Ecflow(CMakePackage):
     maintainers("climbfuji", "AlexanderRichert-NOAA")
 
     # https://confluence.ecmwf.int/download/attachments/8650755/ecFlow-5.8.3-Source.tar.gz?api=v2
```
Can you remove this unnecessary comment please?

srherbener replied:
Sure!

AlexanderRichert-NOAA commented:
Is there some way to automatically infer the standard? In any case it seems like most of the possible variant values don't get used, so it seems like there may be a way to simplify the logic.

srherbener committed:
Fixed pep8 style error in the ecflow package.py script.

climbfuji reviewed (commented):
I assume that ecflow does know about the different cxxstd values and does something with this information? If so, then it's ok imo to add them all. if it's only for the purpose of setting the env variable in `package.py` then I think @AlexanderRichert-NOAA has a good point.

srherbener commented:
> Is there some way to automatically infer the standard? In any case it seems like most of the possible variant values don't get used, so it seems like there may be a way to simplify the logic.

@AlexanderRichert-NOAA and @climbfuji,

I just noticed that this change is missing the setting of the "-std=c++17" compiler option (when cxxstd=17 is specified). That was my original intent. So I would like to make that change and then re-test this. If this works out, are you okay with going with the cxxstd variant?

If not, then I suppose a "boost_no_func_base" variant might be simpler. This would just do the action as it is now, which is to set the BOOST_NO_CXX98_FUNCTION_BASE variable.

srherbener committed:
Removed unnecessary comment

AlexanderRichert-NOAA commented:
@srherbener to me the general notion is to not have distinct variant values that don't "do" anything. So if it can be done automatically or with a simpler variant, great, but if the various values are/can all be used to different effects, then that's fine with me.

srherbener commented:
@AlexanderRichert-NOAA after a more careful look at ecflow, it appears that starting with version 5.8.3 the c++ standard is set in the top level CMakeLists.txt file (ie `set (CMAKE_CXX_STANDARD 17)`). Given this, I think the best option is to infer the need for the BOOST_NO_CXX98_FUNCTION_BASE setting from the version of ecflow. I'll make the changes for this approach. Thanks for the helpful feedback!

climbfuji commented:
> @AlexanderRichert-NOAA after a more careful look at ecflow, it appears that starting with version 5.8.3 the c++ standard is set in the top level CMakeLists.txt file (ie `set (CMAKE_CXX_STANDARD 17)`). Given this, I think the best option is to infer the need for the BOOST_NO_CXX98_FUNCTION_BASE setting from the version of ecflow. I'll make the changes for this approach. Thanks for the helpful feedback!

That sounds great, thanks.

srherbener committed:
Removed cxxstd variant since the ecflow cmake configuration was already specifying
to use the c++17 standard for newer versions. The use of the BOOST_NO_CXX98_FUNCTION_BASE
define is now triggered by the ecflow version.

srherbener committed:
Merge branch 'develop' into feature/ecflow-cxx17

climbfuji reviewed (approved):
I wonder if the new version also fixes the problems we've been seeing building 5.8.3/5.8.4 with gcc@12.2.0 (just a note, not asking for any tests).

climbfuji commented:
@tldahlgren I think this is ready to merge if you are ok with it. Thanks!

tldahlgren reviewed (approved):
Confirmed the version sha256.

climbfuji commented:
Thanks very much @tldahlgren that was super quick!
